src/Features/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs

```diff
@@ -0,0 +1,209 @@
+                        // current set of tasks are done.
+                        // see whether there are new tasks added while we were waiting
+                        break;
+                    }
+
+                    // certain test requires some event queues to be processed
+                    // for waiter tasks to finish such as Dispatcher queue
+                    eventProcessingAction?.Invoke();
+
+                    // in unit test where it uses fake foreground task scheduler such as StaTaskScheduler
+                    // we need to yield for the scheduler to run inlined tasks
+                    await Task.Yield();
+                } while (true);
+            }
+
+            foreach (var task in tasks)
+            {
+                if (task.Exception != null)
+                {
+                    throw task.Exception;
+                }
+            }
+        }
+
+        public bool HasPendingWaiter(params string[] featureNames)
+        {
+            var waiters = GetCandidateWaiters(featureNames);
+            return waiters.Any(w => w.HasPendingWork);
+        }
+
+        /// <summary>
+        /// Get all saved DiagnosticAsyncToken to investigate tests failure easier
+        /// </summary>
+        public List<AsynchronousOperationListener.DiagnosticAsyncToken> GetTokens()
+        {
+            return _singletonListeners.Values.Cast<AsynchronousOperationListener>().Where(l => l.TrackActiveTokens).SelectMany(l => l.ActiveDiagnosticTokens).ToList();
+        }
+
+        private static bool IsEnabled
+        {
+            get
+            {
+                if (!s_enabled.HasValue)
+                {
+                    // if s_enabled has never been set, check environment variable to see whether it should be enabled.
+                    var enabled = Environment.GetEnvironmentVariable("RoslynWaiterEnabled");
+                    s_enabled = string.Equals(enabled, "1", StringComparison.OrdinalIgnoreCase) || string.Equals(enabled, "True", StringComparison.OrdinalIgnoreCase);
+                }
+
+                return s_enabled.Value;
+            }
+        }
+
+        private bool TrackingBehavior
+        {
+            get
+            {
+                if (!_trackingBehavior.HasValue)
+                {
+                    // if _trackingBehavior has never been set, check environment variable to see whether it should be enabled.
+                    var enabled = Environment.GetEnvironmentVariable("RoslynWaiterAsyncTokenTrackingEnabled");
+                    _trackingBehavior = string.Equals(enabled, "1", StringComparison.OrdinalIgnoreCase) || string.Equals(enabled, "True", StringComparison.OrdinalIgnoreCase);
+                }
+
+                return _trackingBehavior.Value;
+            }
+        }
+
+        private IEnumerable<IAsynchronousOperationWaiter> GetCandidateWaiters(string[] featureNames)
+        {
+            if (featureNames == null || featureNames.Length == 0)
```

Comment:
I guess I'd consider null OK but an empty list to be an error...

Reply:
method(params string[] xxx) gives empty array, so need to check empty array.